src/docfx/build/moniker/MonikersProvider.cs

```diff
@@ -51,12 +51,34 @@ public MonikersProvider(Config config, MonikerRangeParser monikerRangeParser)
                 // warn if no intersection of config monikers and file monikers
                 if (intersection.Count == 0)
                 {
-                    error = Errors.NoMonikersIntersection(configMonikerRange, configMonikers, fileMonikerRange, fileMonikers);
+                    error = Errors.NoMonikersIntersection($"No moniker intersection between docfx.yml/docfx.json and file metadata. Config moniker range '{configMonikerRange}' is '{string.Join(',', configMonikers)}', while file moniker range '{fileMonikerRange}' is '{string.Join(',', fileMonikers)}'");
                 }
                 return (error, intersection);
             }
 
             return (error, configMonikers);
         }
+
+        public List<string> GetMonikers(Document file, string rangeString, List<string> fileLevelMonikers, List<Error> errors)
+        {
+            var monikers = new List<string>();
+
+            // Moniker range not defined in docfx.yml/docfx.json,
+            // User should not define it in moniker zone
+            if (fileLevelMonikers.Count == 0)
+            {
+                errors.Add(Errors.MonikerConfigMissing());
+                return new List<string>();
+            }
+
+            var zoneLevelMonikers = file.Docset.MonikerRangeParser.Parse(rangeString);
```

Comment:
>var zoneLevelMonikers = file.Docset.MonikerRangeParser.Parse(rangeString) [](start = 12, length = 73)

`_rules` already caches some result pairs in ` (string monikerRange, List<string> monikers)`. This can also be cached to avoid duplicate parse.

Reply:
yes， we should cache this in parser level

---
In reply to: [233694275](https://github.com/dotnet/docfx/pull/3616#discussion_r233694275) [](ancestors = 233694275)